src/main/java/com/google/devtools/build/lib/analysis/RunfilesSupport.java

```diff
@@ -558,48 +552,12 @@ private static Artifact createRepoMappingManifestAction(
             new RepoMappingManifestAction(
                 ruleContext.getActionOwner(),
                 repoMappingManifest,
-                collectRepoMappings(
-                    Preconditions.checkNotNull(
-                        ruleContext.getTransitivePackagesForRunfileRepoMappingManifest()),
-                    runfiles),
+                ruleContext.getTransitivePackagesForRunfileRepoMappingManifest(),
+                runfiles.getAllArtifacts(),
```

Comment:
FYI: getAllArtifacts() does a partial flattening -- symlinks and root symlinks are still flattened. But those those are usually small, so still a big win.

Reply:
Do you think it would be worth improving that further? Can runfiles be serialized? Then that could make for a better action constructor argument.

Reply:
I think so? SourceManifestAction, for example, accepts a Runfiles object directly.

I wouldn't expect a large gain in practice. The symlinks and root_symlinks are usually small. Runfiles.getAllArtifacts() is called in a few different places already.

Reply:
Thanks for your comment, it led me to discover that the repo mapping manifest didn't handle symlinks correctly: https://github.com/bazelbuild/bazel/pull/18381